src/blockstore/hashmap_db.zig

```diff
@@ -0,0 +1,237 @@
+                    },
+                }
+            }
+        }
+
+        pub const WriteBatch = struct {
+            allocator: Allocator,
+            instructions: std.ArrayListUnmanaged(Instruction),
+
+            const Instruction = union(enum) {
+                put: struct { usize, []const u8, []const u8 },
+                delete: struct { usize, []const u8 },
+            };
+
+            fn deinit(self: WriteBatch) void {
+                self.instructions.deinit(self.allocator);
+            }
+
+            pub fn put(
+                self: *WriteBatch,
+                comptime cf: ColumnFamily,
+                key: cf.Key,
+                value: cf.Value,
+            ) Allocator.Error!void {
+                const k_bytes = try serializeAlloc(self.allocator, key);
+                errdefer self.allocator.free(k_bytes);
+                const v_bytes = try serializeAlloc(self.allocator, value);
+                errdefer self.allocator.free(v_bytes);
+                return try self.instructions.append(
+                    self.allocator,
+                    .{ .put = .{ cf.find(column_families), k_bytes, v_bytes } },
+                );
+            }
+
+            pub fn delete(
+                self: *WriteBatch,
+                comptime cf: ColumnFamily,
+                key: cf.Key,
+            ) Allocator.Error!void {
+                const k_bytes = try serializeAlloc(self.allocator, key);
+                errdefer self.allocator.free(k_bytes);
+                return try self.instructions.append(
+                    self.allocator,
+                    .{ .delete = .{ cf.find(column_families), k_bytes } },
+                );
+            }
+        };
+    };
+}
+
+const SharedHashMap = struct {
+    allocator: Allocator,
+    map: std.StringHashMapUnmanaged([]const u8) = .{},
+    lock: DefaultRwLock = .{},
+
+    const Self = @This();
+
+    fn init(allocator: Allocator) Allocator.Error!Self {
+        return .{ .allocator = allocator };
+    }
+
+    pub fn deinit(self: *Self) void {
+        var iter = self.map.iterator();
+        while (iter.next()) |entry| {
+            self.allocator.free(entry.key_ptr.*);
+            self.allocator.free(entry.value_ptr.*);
+        }
+        self.map.deinit(self.allocator);
+    }
+
+    pub fn put(self: *Self, key: []const u8, value: []const u8) Allocator.Error!void {
+        self.lock.lock();
+        defer self.lock.unlock();
+        try self.map.put(self.allocator, key, value);
+    }
+
+    /// Only call this while holding the lock
+    pub fn getPreLocked(self: *Self, key: []const u8) ?[]const u8 {
+        return self.map.get(key) orelse return null;
+    }
+
+    pub fn delete(self: *Self, liberator: Allocator, key_: []const u8) void {
+        const key, const value = lock: {
+            self.lock.lock();
+            defer self.lock.unlock();
+            const entry = self.map.getEntry(key_) orelse return;
+            const key = entry.key_ptr.*;
```

Comment:
why do we need this? - cant we just return `key_` ? -- would also help with naming since we could remove the suffix _